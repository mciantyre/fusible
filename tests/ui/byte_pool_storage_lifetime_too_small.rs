// SPDX-License-Identifier: MPL-2.0
// SPDX-FileCopyrightText: Copyright 2024 Ian McIntyre

use fusible::byte_pool::{BytePool, BytePoolOptions};

/// The pool outlives the storage. When this happens, the pool holds
/// a dangling reference to the storage. The compiler rejects this
/// ordering.
fn bi() {
    let pool = core::pin::pin!(BytePool::context());
    let mut storage = [0u8; 1024];
    BytePool::create(pool.as_ref(), &mut storage, &{
        let mut opts = BytePoolOptions::default();
        opts.name = Some(c"name");
        opts
    })
    .unwrap();
}

/// Multiple create calls would appear to compile (though they'll
/// fail at runtime, which isn't tested here). The compiler rejects
/// call that uses the storage with a smaller lifetime. This checks
/// our understanding of how the compiler infers the pool's inner
/// lifetime.
fn bim() {
    let mut upper = [0u8; 512];
    let pool = core::pin::pin!(BytePool::context());
    let mut lower = [0u8; 512];

    BytePool::create(pool.as_ref(), &mut upper, &Default::default()).unwrap(); // OK: upper lifetime greater than pool.
    BytePool::create(pool.as_ref(), &mut lower, &Default::default()).unwrap(); // ERROR: lower lifetime less than pool.
}

/// Multiple create calls would appear to compile (though they'll
/// fail at runtime, which isn't tested here). The compiler rejects
/// call that uses the storage with a smaller lifetime. This checks
/// our understanding of how the compiler infers the pool's inner
/// lifetime.
fn bap() {
    let mut upper = [0u8; 512];
    let pool = core::pin::pin!(BytePool::context());
    let mut lower = [0u8; 512];

    BytePool::create(pool.as_ref(), &mut lower, &Default::default()).unwrap(); // ERROR: lower lifetime less than pool.
    BytePool::create(pool.as_ref(), &mut upper, &Default::default()).unwrap(); // OK: upper lifetime greater than pool.
}

fn main() {
    bi();
    bim();
    bap();
}
