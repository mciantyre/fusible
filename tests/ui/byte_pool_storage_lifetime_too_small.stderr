error[E0597]: `storage` does not live long enough
  --> tests/ui/byte_pool_storage_lifetime_too_small.rs:12:37
   |
11 |     let mut storage = [0u8; 1024];
   |         ----------- binding `storage` declared here
12 |     BytePool::create(pool.as_ref(), &mut storage, &{
   |                                     ^^^^^^^^^^^^ borrowed value does not live long enough
...
18 | }
   | -
   | |
   | `storage` dropped here while still borrowed
   | borrow might be used here, when `pinned` is dropped and runs the `Drop` code for type `BytePoolContext`
   |
   = note: values in a scope are dropped in the opposite order they are defined

error[E0597]: `lower` does not live long enough
  --> tests/ui/byte_pool_storage_lifetime_too_small.rs:31:37
   |
28 |     let mut lower = [0u8; 512];
   |         --------- binding `lower` declared here
...
31 |     BytePool::create(pool.as_ref(), &mut lower, &Default::default()).unwrap(); // ERROR: lower lifetime less than pool.
   |                                     ^^^^^^^^^^ borrowed value does not live long enough
32 | }
   | -
   | |
   | `lower` dropped here while still borrowed
   | borrow might be used here, when `pinned` is dropped and runs the `Drop` code for type `BytePoolContext`
   |
   = note: values in a scope are dropped in the opposite order they are defined

error[E0597]: `lower` does not live long enough
  --> tests/ui/byte_pool_storage_lifetime_too_small.rs:44:37
   |
42 |     let mut lower = [0u8; 512];
   |         --------- binding `lower` declared here
43 |
44 |     BytePool::create(pool.as_ref(), &mut lower, &Default::default()).unwrap(); // ERROR: lower lifetime less than pool.
   |                                     ^^^^^^^^^^ borrowed value does not live long enough
45 |     BytePool::create(pool.as_ref(), &mut upper, &Default::default()).unwrap(); // OK: upper lifetime greater than pool.
46 | }
   | -
   | |
   | `lower` dropped here while still borrowed
   | borrow might be used here, when `pinned` is dropped and runs the `Drop` code for type `BytePoolContext`
   |
   = note: values in a scope are dropped in the opposite order they are defined
